### PR TITLE
UnmarshalText precedence with tests

### PR DESCRIPTION
**Pull request by Azuka**
Pulls in previous work done on #58 and #64 with added tests.

Slices of pointers appear to fail with this though. I'm not sure how to fix, if anyone has suggestions.

### CODE REVIEW

File: decoder.go [hunk 1]
```diff
@@ -275,31 +302,36 @@ func (d *Decoder) decode(v reflect.Value, path string, parts []pathPart, values
...
+		return u, ok
+	}
+
+	// if v is []T or *[]T create new T
+	t := v.Type()
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+	if t.Kind() == reflect.Slice {
+		t = t.Elem()
```
Comment (kisielk) [inline decoder.go]: I think you need to check for `t.Kind() == reflect.Ptr` here as well

File: decoder.go [hunk 2]
```diff
@@ -188,6 +188,7 @@ func (d *Decoder) decode(v reflect.Value, path string, parts []pathPart, values
 	if conv == nil && t.Kind() == reflect.Slice {
 		var items []reflect.Value
 		elemT := t.Elem()
+		//elemP := elemT
```
Comment (kisielk) [inline decoder.go]: Unused code

Comment (Azuka) [inline decoder.go]: Didn't notice that. I cherry-picked the previous commits as-is. I'll remove this now.

File: decoder.go (hunk 1 above)
Comment (Azuka) [inline decoder.go]: Thanks. I'll try that.

Comment (Azuka) [inline decoder.go]: Not sure I understand reflection well enough in go.

Replacing `t = t.Elem()` above with:

```go
if t = t.Elem(); t.Kind() == reflect.Ptr {
    t = t.Elem()
}
```
I get `panic: reflect.Set: value of type schema.rudeBool is not assignable to type *schema.rudeBool`

and with:

```go
if t = t.Elem(); t.Kind() == reflect.Ptr {
    t = reflect.PtrTo(t.Elem())
}
```
I get nil which fails the test for the slice pointer array

**Comment by Azuka** [comment]
*Should* be ready to go. Open to suggestions on how to clean up the code, but this works.

File: decoder.go [hunk 3]
```diff
@@ -275,29 +306,39 @@ func (d *Decoder) decode(v reflect.Value, path string, parts []pathPart, values
...
-			if u, ok := v.Interface().(encoding.TextUnmarshaler); ok {
-				if err := u.UnmarshalText([]byte(val)); err != nil {
-					return ConversionError{
-						Key:   path,
-						Type:  t,
-						Index: -1,
-						Err:   err,
-					}
-				}
+func isTextUnmarshaler(v reflect.Value) (encoding.TextUnmarshaler, bool, bool) {
```
Comment (kisielk) [inline decoder.go]: I think the return values from this function are becoming confusing now. Maybe it should just return a struct?

```
struct unmarshaller {
    U encoding.TextUnmarhaler
    IsPtr bool
    IsValid bool
}
```

**Comment by Azuka** [comment]
@kisielk, updated.

**Comment by kisielk** [comment]
Looks good to me!

**Comment by HasMatthew** [comment]
This change breaks my existing use case for defining a custom `time.Time` decoder (which implements the `encoding.UnmarshalText` interface).

Is there a reason why `Decode` would ignore an explicitly registered converter for a type that happens to implement the `encoding.UnmarshalText` interface?

**Comment by kisielk** [comment]
@HasMatthew that's a good point, I definitely missed that use case in the review.

I can see why the code was written the way it was though, since the converter cache is pre-populated with all the default converters then it needed to check for the `TextUnmarshaler` interface first or else it would not be possible to override them.

Really the code needs to have 3 levels of converter lookup:

1. Explicitly registered converter by the user.
2. `TextUnmarshaler` interface.
3. Default converter for base type.

Currently 1 and 3 are lumped together and happen after 2.

**Comment by HasMatthew** [comment]
Yeah, it seems you need a separate map for user registered converters and default converters, which would allow you to prioritize the deserialization methods that you described.

**Pull request merged**